File: app/src/main/java/igrek/forceawaken/service/task/AwakeTaskService.java
package igrek.forceawaken.service.task;

import android.app.Activity;

import java.util.LinkedList;
import java.util.Random;

import igrek.forceawaken.domain.task.AnswerAgainTask;
import igrek.forceawaken.domain.task.AwakeTask;
import igrek.forceawaken.domain.task.LuckyTask;
import igrek.forceawaken.domain.task.MathTask;
import igrek.forceawaken.logger.Logger;

public class AwakeTaskService {
	
	private Activity activity;
	private Logger logger;
	private LinkedList<AwakeTask> registeredTasks = new LinkedList<>();
	private Random random = new Random();
	
	public AwakeTaskService(Activity activity, Logger logger) {
		this.activity = activity;
		this.logger = logger;
		logger.dupa();
		enableTasks();
	}
	
	private void enableTasks() {
		registeredTasks.add(new LuckyTask());
		registeredTasks.add(new MathTask());
		registeredTasks.add(new AnswerAgainTask());
	}
	
	private double sumTasksProbability() {
		double sum = 0;
		for (AwakeTask registeredTask : registeredTasks) {
			sum += registeredTask.getProbabilityWeight();
		}
		return sum;
	}
	
	public AwakeTask getRandomTask() {
		// not uniform random
		double sumP = sumTasksProbability();
		double offset = random.nextDouble() * sumP;
		double sum = 0;
		for (AwakeTask registeredTask : registeredTasks) {
			sum += registeredTask.getProbabilityWeight();
			if (sum > offset)
				return registeredTask.getInstance();
		}
		return registeredTasks.getLast();
	}
	
}

File: app/src/main/java/igrek/forceawaken/task/AwakeTaskService.kt
package igrek.forceawaken.task

import android.app.Activity
import igrek.forceawaken.info.logger.Logger
import igrek.forceawaken.info.logger.LoggerFactory
import igrek.forceawaken.inject.LazyExtractor
import igrek.forceawaken.inject.LazyInject
import igrek.forceawaken.inject.appFactory
import java.util.*

class AwakeTaskService(
        activity: LazyInject<Activity> = appFactory.activity,
) {
    private val activity by LazyExtractor(activity)

    private val logger: Logger = LoggerFactory.logger

    private val registeredTasks: LinkedList<AwakeTask> = LinkedList<AwakeTask>()
    private val random = Random()

    init {
        enableTasks()
    }

    private fun enableTasks() {
        registeredTasks.add(LuckyTask())
        registeredTasks.add(AnswerAgainTask())
    }

    private fun sumTasksProbability(): Double {
        var sum = 0.0
        for (registeredTask in registeredTasks) {
            sum += registeredTask.probabilityWeight
        }
        return sum
    }

    // not uniform random
    val randomTask: AwakeTask
        get() {
            // not uniform random
            val sumP = sumTasksProbability()
            val offset = random.nextDouble() * sumP
            var sum = 0.0
            for (registeredTask in registeredTasks) {
                sum += registeredTask.probabilityWeight
                if (sum > offset) return registeredTask.instance
            }
            return registeredTasks.last
        }

}